src/lib/storage/table.cpp

```diff
@@ -16,54 +16,66 @@
 #include "utils/assert.hpp"
 
 namespace opossum {
+Table::Table(uint32_t chunk_size) : _max_chunk_size(chunk_size) { _chunks.push_back(std::make_shared<Chunk>()); }
 
 void Table::add_column(const std::string& name, const std::string& type) {
-  // Implementation goes here
+  DebugAssert(row_count() == 0, "The table already contains data so adding a column is not possible.");
+
+  _column_names.push_back(name);
+  _column_types.push_back(type);
+
+  auto segment = make_shared_by_data_type<BaseSegment, ValueSegment>(type);
+
+  _chunks.front()->add_segment(segment);
 }
 
 void Table::append(std::vector<AllTypeVariant> values) {
-  // Implementation goes here
-}
+  DebugAssert(values.size() == column_count(), "The number of values doesn't match number of columns.");
 
-uint16_t Table::column_count() const {
-  // Implementation goes here
-  return 0;
+  if (_chunks.back()->size() >= _max_chunk_size) {
+    _chunks.push_back(std::make_shared<Chunk>());
+
+    for (const auto& _column_type : _column_types) {
+      auto segment = make_shared_by_data_type<BaseSegment, ValueSegment>(_column_type);
+
+      _chunks.back()->add_segment(segment);
+    }
+  }
+
+  _chunks.back()->append(values);
 }
 
+uint16_t Table::column_count() const { return _chunks.front()->column_count(); }
+
 uint64_t Table::row_count() const {
-  // Implementation goes here
-  return 0;
+  uint64_t number_of_rows = 0;
```

Comment:
counting rows now happens in O(n).
If we had an additional member variable _lastChunk = 0, which we always increment, when a new chunk is added, we could compute number of rows in O(1).
```
 return _lastChunk * _max_chunk_size + _chunks[_lastChunk]->size();
```

Reply:
I changed it to this version without the new variable

`return (_chunks.size() - 1) * _max_chunk_size + _chunks.back()->size();`

Reply:
Oh that's even better! :)